File: auth-service/src/auth/auth.service.ts
import { BadRequestException, Injectable, Inject, NotFoundException, UnauthorizedException, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { JwtService } from '@nestjs/jwt';
import { ClientProxy } from '@nestjs/microservices';
import * as bcrypt from 'bcrypt';
import { User, UserRole } from './entities/user.entity';

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    @InjectRepository(User) private usersRepository: Repository<User>,
    private jwtService: JwtService,
    @Inject('EVENT_EXCHANGE') private publisher: ClientProxy,
  ) {}

  async onModuleInit() {
    await this.publisher.connect();
  }

  async login(loginDto: any) {
    const { email, password } = loginDto;
    const user = await this.usersRepository.findOne({ where: { email } });
    if (!user || !user.isActive) {
      throw new UnauthorizedException('Invalid credentials or inactive user');
    }

    if (!user.password) {
      throw new UnauthorizedException('Please reset your password before logging in');
    }

    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const payload = { sub: user.id, role: user.role };
    return {
      access_token: this.jwtService.sign(payload, { expiresIn: '1h' }),
    };
  }

  async recoverPassword(email: string) {
    const user = await this.usersRepository.findOne({ where: { email } });
    if (!user || !user.isActive) {
      throw new NotFoundException('User not found or inactive');
    }

    const payload = { sub: user.id, type: 'RESET_PASSWORD' };
    const resetToken = this.jwtService.sign(payload, { expiresIn: '15m' });

    // Emit event for notification
    this.publisher.emit('user.recovered', {
      email: user.email,
      token: resetToken,
    });
    this.logger.log(`Emitted user.recovered for ${user.email}`);

    return { message: 'Recovery email simulated' };
  }

  async resetPassword(resetDto: any) {
    const { token, newPassword } = resetDto;

    try {
      const payload = this.jwtService.verify(token);
      if (payload.type !== 'RESET_PASSWORD') {
        throw new BadRequestException('Invalid token type');
      }

      const user = await this.usersRepository.findOne({ where: { id: payload.sub } });
      if (!user) throw new NotFoundException('User not found');

      const salt = await bcrypt.genSalt();
      user.password = await bcrypt.hash(newPassword, salt);
      await this.usersRepository.save(user);

      return { message: 'Password updated successfully' };
    } catch (e) {
      throw new BadRequestException('Invalid or expired token');
    }
  }

  async handleEmployeeCreated(data: any) {
    // Check if user exists
    let user = await this.usersRepository.findOne({ where: { email: data.email } });
    if (!user) {
      user = this.usersRepository.create({
         email: data.email,
         role: UserRole.USER,
         isActive: true,
      });
      await this.usersRepository.save(user);
    }
    
    // Generate reset token as they have no password
    const payload = { sub: user.id, type: 'RESET_PASSWORD' };
    const resetToken = this.jwtService.sign(payload, { expiresIn: '2h' });

    this.publisher.emit('user.created', {
      email: user.email,
      token: resetToken,
    });
    this.logger.log(`Created user ${user.email} from employee and emitted user.created`);
  }

  async handleEmployeeDeleted(data: any) {
    const user = await this.usersRepository.findOne({ where: { email: data.email } });
    if (user) {
      user.isActive = false;
      await this.usersRepository.save(user);
      this.logger.log(`Deactivated user ${user.email} from employee deleted`);
    }
  }
}

File: auth-service/src/main/java/com/empresa/auth_service/service/AuthService.java
package com.empresa.auth_service.service;

import com.empresa.auth_service.dto.LoginDto;
import com.empresa.auth_service.dto.ResetDto;
import com.empresa.auth_service.model.User;
import com.empresa.auth_service.repository.UserRepository;
import com.empresa.auth_service.security.JwtUtil;
import io.jsonwebtoken.Claims;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public class AuthService {

    private static final Logger logger = LoggerFactory.getLogger(AuthService.class);

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private JwtUtil jwtUtil;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private RabbitTemplate rabbitTemplate;

    public Map<String, String> login(LoginDto loginDto) {
        User user = userRepository.findByEmail(loginDto.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid credentials or inactive user"));

        if (!user.isActive()) {
            throw new RuntimeException("Invalid credentials or inactive user");
        }

        if (user.getPassword() == null || user.getPassword().isEmpty()) {
            throw new RuntimeException("Please reset your password before logging in");
        }

        if (!passwordEncoder.matches(loginDto.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid credentials");
        }

        Map<String, Object> claims = new HashMap<>();
        claims.put("role", user.getRole());

        String token = jwtUtil.generateToken(user.getId(), claims, 3600000); // 1h

        Map<String, String> response = new HashMap<>();
        response.put("access_token", token);
        return response;
    }

    public Map<String, String> recoverPassword(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found or inactive"));

        if (!user.isActive()) throw new RuntimeException("User not found or inactive");

        Map<String, Object> claims = new HashMap<>();
        claims.put("type", "RESET_PASSWORD");
        String resetToken = jwtUtil.generateToken(user.getId(), claims, 900000); // 15m

        // Emit event user.recovered
        String payload = "{\"pattern\":\"user.recovered\",\"data\":{\"email\":\"" + email + "\",\"token\":\"" + resetToken + "\"}}";
        rabbitTemplate.convertAndSend("employees_exchange", "", payload);
        logger.info("Emitted user.recovered for " + email);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Recovery email simulated");
        return response;
    }

    public Map<String, String> resetPassword(ResetDto resetDto) {
        try {
            Claims claims = jwtUtil.validateToken(resetDto.getToken());
            if (!"RESET_PASSWORD".equals(claims.get("type"))) {
                throw new RuntimeException("Invalid token type");
            }

            User user = userRepository.findById(claims.getSubject())
                    .orElseThrow(() -> new RuntimeException("User not found"));

            user.setPassword(passwordEncoder.encode(resetDto.getNewPassword()));
            userRepository.save(user);

            Map<String, String> response = new HashMap<>();
            response.put("message", "Password updated successfully");
            return response;
        } catch (Exception e) {
            throw new RuntimeException("Invalid or expired token");
        }
    }

    @RabbitListener(queues = "auth_queue")
    public void receiveMessage(String message) {
        logger.info("Received event: " + message);
        try {
            // Very naive JSON parsing without Jackson overhead just for event pattern
            if (message.contains("\"pattern\":\"employee.created\"")) {
                String email = extractJsonStringValue(message, "email");
                handleEmployeeCreated(email);
            } else if (message.contains("\"pattern\":\"employee.deleted\"")) {
                String email = extractJsonStringValue(message, "email");
                handleEmployeeDeleted(email);
            }
        } catch (Exception e) {
            logger.error("Error processing event", e);
        }
    }

    private void handleEmployeeCreated(String email) {
        Optional<User> existing = userRepository.findByEmail(email);
        User user;
        if (existing.isPresent()) {
            user = existing.get();
        } else {
            user = new User();
            user.setEmail(email);
            user.setRole("USER");
            user.setActive(true);
            user = userRepository.save(user);
        }

        Map<String, Object> claims = new HashMap<>();
        claims.put("type", "RESET_PASSWORD");
        String resetToken = jwtUtil.generateToken(user.getId(), claims, 7200000); // 2h

        String payload = "{\"pattern\":\"user.created\",\"data\":{\"email\":\"" + email + "\",\"token\":\"" + resetToken + "\"}}";
        rabbitTemplate.convertAndSend("employees_exchange", "", payload);
        logger.info("Created user " + email + " from employee and emitted user.created");
    }

    private void handleEmployeeDeleted(String email) {
        Optional<User> existing = userRepository.findByEmail(email);
        existing.ifPresent(user -> {
            user.setActive(false);
            userRepository.save(user);
            logger.info("Deactivated user " + email + " from employee deleted");
        });
    }

    private String extractJsonStringValue(String json, String key) {
        String search = "\"" + key + "\":\"";
        int start = json.indexOf(search);
        if (start == -1) return null;
        start += search.length();
        int end = json.indexOf("\"", start);
        return json.substring(start, end);
    }
}
